File: domain/src/testFixtures/java/com/study/platform/support/fake/FakeKafkaMessagePublisher.java
package com.study.platform.support.fake;

import com.study.platform.global.kafka.KafkaMessagePublisher;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.CompletableFuture;

public class FakeKafkaMessagePublisher implements KafkaMessagePublisher {

    public record PublishedMessage(String topic, String key, String payload) {}

    private final List<PublishedMessage> published = new ArrayList<>();
    private boolean shouldFail = false;

    @Override
    public CompletableFuture<Void> publish(String topic, String key, String payload) {
        if (shouldFail) {
            return CompletableFuture.failedFuture(new RuntimeException("Kafka 연결 실패"));
        }
        published.add(new PublishedMessage(topic, key, payload));
        return CompletableFuture.completedFuture(null);
    }

    public void willFail() {
        shouldFail = true;
    }

    public boolean wasPublishedTo(String topic) {
        return published.stream().anyMatch(m -> m.topic().equals(topic));
    }

    public List<String> getPublishedTopics() {
        return published.stream().map(PublishedMessage::topic).toList();
    }

    public List<PublishedMessage> getPublished() {
        return published;
    }
}

File: domain/src/testFixtures/kotlin/com/study/platform/support/fake/FakeKafkaMessagePublisher.kt
package com.study.platform.support.fake

import com.study.platform.global.kafka.KafkaMessagePublisher
import java.util.concurrent.CompletableFuture

class FakeKafkaMessagePublisher : KafkaMessagePublisher {

    data class PublishedMessage(val topic: String, val key: String?, val payload: String)

    private val published: MutableList<PublishedMessage> = ArrayList()
    private var shouldFail = false

    override fun publish(topic: String, key: String?, payload: String): CompletableFuture<Void> {
        if (shouldFail) {
            return CompletableFuture.failedFuture(RuntimeException("Kafka 연결 실패"))
        }
        published.add(PublishedMessage(topic, key, payload))
        return CompletableFuture.completedFuture(null)
    }

    fun willFail() {
        shouldFail = true
    }

    fun wasPublishedTo(topic: String): Boolean =
        published.any { m -> m.topic == topic }

    fun getPublishedTopics(): List<String> =
        published.map { it.topic }

    fun getPublished(): List<PublishedMessage> = published
}
